WordPress/gutenberg
# Block Hooks: Add controller to expose hooked blocks for a given anchor block

_More details to come!_

## What?
Add a REST API controller for a new `/hooked-blocks` route that exposes hooked blocks for a given anchor block, when used in a certain context (e.g. within a template, a pattern, or a navigation menu).

This allows obtaining a list of blocks that are hooked to the given anchor within the current context. This takes blocks into account that are added (or removed) via the `hooked_block_types` filter -- unlike the `block-types` endpoint, which is only aware of the context-agnostic `blockHooks` field in `block.json`.

## Why?
See https://core.trac.wordpress.org/ticket/59574 and https://github.com/WordPress/gutenberg/pull/58553.

## How?
TBD

## Testing Instructions
TBD, but largely: Add a hooked block via the `hooked_block_types` filter, and verify that the endpoint includes it.
E.g. in the browser console:
```js
let result = await wp.apiFetch( { path: '/wp/v2/hooked-blocks/core/post-content?entity=template&id=twentytwentythree/single' } );
```


## Screenshots or screencast 

## TODO

_Work is largely inspired by the [`block-types` endpoint](https://developer.wordpress.org/rest-api/reference/block-types/)_

- [ ] Test with conditionally added blocks (i.e. via `hooked_block_types` filter) and verify that they're present in the endpoint's response; add unit test coverage for that.
- [ ] Build `$context` from query args, pass to `hooked_block_types` filter.
  - [x] Template Part
  - [ ] Pattern?
  - [x] Navigation 
- [ ] Add schema
- [x] Get the `/hooked-blocks/` "root" to work.
- [ ] ~Filter fields (`after`, `before`, ... )~
- [ ] Re-write `getHookedBlocks` selector (introduced by #58553) to use endpoint as source of truth.
